apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs

```diff
@@ -603,23 +604,30 @@ private object ConvertToClrTypeImpl(Value wireValue, System.Type targetClrType,
                         return null;
                     case Value.KindOneofCase.StringValue:
                         return SpannerNumeric.Parse(wireValue.StringValue);
+                    case Value.KindOneofCase.NumberValue:
+                        return SpannerNumeric.FromDecimal(Convert.ToDecimal(wireValue.NumberValue), LossOfPrecisionHandling.Truncate);
                     default:
                         throw new InvalidOperationException(
                             $"Invalid Type conversion from {wireValue.KindCase} to {targetClrType.FullName}");
                 }
             }
             if (targetClrType == typeof(PgNumeric))
             {
-                if (TypeCode != TypeCode.Numeric || TypeAnnotationCode != TypeAnnotationCode.PgNumeric)
+                bool isPgNumeric = TypeAnnotationCode == TypeAnnotationCode.PgNumeric;
+                bool isFloat64 = TypeCode == TypeCode.Float64;
+
+                if (!isPgNumeric && !isFloat64)
```

Comment:
Should this be `||` instead of `&&`? (At the moment, we wouldn't fail if it's a string with a PgNumeric type annotation code for example. That might never be feasible, but it feels like it's logically `||`.)

Reply:
We need `&&` here as PgNumeric value by default comes as string and while reading a Float64 value as PgNumeric, we need to read the NumberValue :(

Reply:
I'm afraid I don't understand - and in particular, this *is* a change to behavior.
Was there a bug before?

As a concrete example: we *won't* throw an exception here if the user has requested a `PgNumeric` conversion for a `TypeCode.Float64` value in a regular Spanner database... that seems wrong, given the error message. Surely we want to throw an exception if *either* we're not in Postgres *or* we don't have a numeric value.

Reply:
Okay, it sounds like what you *really* want is something saying "is the type code string or float64" - although I may have misunderstood the relationship between the type code and the Value kind here.

Fundamentally this code needs to be much clearer about what's valid to convert into PgNumeric...